stdlib/public/core/Unicode2.swift

```diff
@@ -0,0 +1,1129 @@
+//===--- Unicode2.swift ---------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// The result of parsing input from a collection.
+///
+/// - Parameter T: the result of a successful parse
+/// - Parameter Index: the index type of the collection being transcoded.
+public enum ParseResult<T, Index> { // FIXME: RENAME THIS!
+/// Indicates valid input was recognized.
+///
+/// `resumptionPoint` is the end of the parsed region
+case valid(T, resumptionPoint: Index)  // FIXME: should these be reordered?
+
+/// Indicates invalid input was recognized.
+///
+/// `resumptionPoint` is the next position at which to continue parsing after
+/// the invalid input is repaired.
+case error(resumptionPoint: Index)
+
+/// Indicates that there was no more input to consume.
+case emptyInput
+
+  /// If any input was consumed, the point from which to continue parsing.
+  var resumptionPoint: Index? {
+    switch self {
+    case .valid(_,let r): return r
+    case .error(let r): return r
+    case .emptyInput: return nil
+    }
+  }
+}
+
+/// An encoding for text with UnicodeScalar as a common currency type
+public protocol UnicodeEncoding {  
+  // FIXME: a single scalar might not be the most efficient buffer to use here.
+  // SIMD instructions can be used to decode UTF-8 much more efficiently, which
+  // could result in processing up to 16 code units at a time.
+  /// The maximum number of code units in an encoded unicoded scalar value
+  static var maxLengthOfEncodedScalar: UInt { get }
+  
+  /// A type that can represent a single UnicodeScalar as it is encoded in this
+  /// encoding.
+  associatedtype EncodedScalar : EncodedScalarProtocol
+  // where Iterator.Element : UnsignedInteger
+  
+  associatedtype CodeUnit : UnsignedInteger
+    where CodeUnit == EncodedScalar.Iterator.Element
```

Comment:
@DougGregor check it out.  I guess I ought to take the opportunity to do the other ones!